helix-term/src/commands/typed.rs

```diff
@@ -1410,47 +1410,63 @@ fn tutor(
     Ok(())
 }
 
+fn abort_goto_line_number_preview(cx: &mut compositor::Context) {
+    if let Some(last_selection) = cx.editor.last_selection.take() {
+        let scrolloff = cx.editor.config().scrolloff;
+
+        let (view, doc) = current!(cx.editor);
+        doc.set_selection(view.id, last_selection.clone());
+        view.ensure_cursor_in_view(doc, scrolloff);
+    }
+}
+
+fn update_goto_line_number_preview(
+    cx: &mut compositor::Context,
+    args: &[Cow<str>],
+) -> anyhow::Result<()> {
+    cx.editor.last_selection.get_or_insert_with(|| {
+        let (view, doc) = current!(cx.editor);
+        doc.selection(view.id).clone()
+    });
+
+    let scrolloff = cx.editor.config().scrolloff;
+    let line = args[0].parse::<usize>()?;
+    goto_line_without_jumplist(cx.editor, NonZeroUsize::new(line));
+
+    let (view, doc) = current!(cx.editor);
+    view.ensure_cursor_in_view(doc, scrolloff);
+
+    Ok(())
+}
+
 pub(super) fn goto_line_number(
     cx: &mut compositor::Context,
     args: &[Cow<str>],
     event: PromptEvent,
 ) -> anyhow::Result<()> {
     match event {
-        PromptEvent::Abort => {
-            if let Some(line_number) = cx.editor.last_line_number {
-                goto_line_impl(cx.editor, NonZeroUsize::new(line_number));
-                let (view, doc) = current!(cx.editor);
-                view.ensure_cursor_in_view(doc, line_number);
-                cx.editor.last_line_number = None;
-            }
-            return Ok(());
-        }
+        PromptEvent::Abort => abort_goto_line_number_preview(cx),
         PromptEvent::Validate => {
             ensure!(!args.is_empty(), "Line number required");
-            cx.editor.last_line_number = None;
-        }
-        PromptEvent::Update => {
-            if args.is_empty() {
-                if let Some(line_number) = cx.editor.last_line_number {
-                    // When a user hits backspace and there are no numbers left,
-                    // we can bring them back to their original line
-                    goto_line_impl(cx.editor, NonZeroUsize::new(line_number));
-                    let (view, doc) = current!(cx.editor);
-                    view.ensure_cursor_in_view(doc, line_number);
-                    cx.editor.last_line_number = None;
-                }
-                return Ok(());
+
+            // If we are invoked directly via a keybinding, Validate is
+            // sent without any prior Update events. Ensure the cursor
+            // is moved to the appropriate location.
+            update_goto_line_number_preview(cx, args)?;
+
+            if let Some(last_selection) = cx.editor.last_selection.take() {
```

Comment:
Now that `last_selection` is always populated in `update_goto_line_number_preview` you  should call `unwrap` here instead of `if let`. If `last_selection` somehow became `None` in the meantime that is a bug and therefore correct to panic rather than silently ignoring the problem. Fail hard and fail fast...

Reply:
> Fail hard and fail fast...

Interesting. I did consider this code path and had the opposite instinct. In the event of a code change that results in a bug I didn't want to crash and potentially lose a user's unsaved editor changes just because a jumplist entry couldn't be set. Is this approach to error handling standard for the project, and if so, documented for contributors?

Alternatively, we could use either a debug_assert! or else conditional with logging if you're concerned about detecting such a bug.

Reply:
you can `bail` too in the else case if you are super concerned about the `else` case causing data loss but  If you  look throughout the codebase, `unwrap` and `expect` are used liberally for impossible code-paths. There was some discussion about adding a panic handler that backups  unsaved work. Rust autoamtically panics on many common programmar bugs (out of bounds indexing, etc.) so that problem is not specific to this one case.

Right now the code looks like it's perfectly valid for there to be no `last_selection` when it should really be impossible. That is an anitpattern to me.. Using unwrap for code-paths that should never happens is a pretty common idiom in rust and something that makes code easier to read (which we care about) and makes it easier to find bugs.

Reply:
No problem, happy to be consistent with the rest of the project. Pushed a fix for that and the clippy warnings.